frontend/src/metabase-lib/parameters/utils/parameter-fields.ts

```diff
@@ -7,7 +7,11 @@ import type {
 export const hasFields = (
   parameter: UiParameter,
 ): parameter is FieldFilterUiParameter => {
-  return (parameter as FieldFilterUiParameter).fields != null;
+  return (
+    "fields" in parameter &&
+    parameter.fields != null &&
+    parameter.fields.length > 0
```

Comment:
Good point, I think you're right there. I'd prefer to refactor that in a separate PR though immediately after this one is approved.

Reply:
I've created a PR to do this refactoring [here](https://github.com/metabase/metabase/pull/35086/files#diff-71b7995e3b36068e5ba6e07964575b1f07f14f8bcb15d9bee355f7078d9e2311R7)

Reply:
@kamilmielnik I tried the refactoring and it was rejected on the basis that these sort of types are difficult to enforce with the type system. I think the principle that Alex is advocating is if you can't easily enforce a constraint, don't write it down using types. See https://github.com/metabase/metabase/pull/35086

Reply:
The problem I was commenting on is that a type guard contains application logic:
- `parameter.fields.length > 0` does not belong in a type guard
- another symptom is that the type guard is called `hasFields` (instead of `isX`, where X is name of the type)

> it'd be better to introduce a separate, boolean-returning function that has this condition.
And then this type guard should be renamed to `isFieldFilterUiParameter`, while that boolean-returning function could be named `hasFields`.

This suggestion is still valid.

To illustrate the suggestion with code:
```tsx
export const isFieldFilterUiParameter = (
  parameter: UiParameter,
): parameter is FieldFilterUiParameter => {
  return "fields" in parameter && parameter.fields != null; // this could be improved though
};

export const hasFields = (parameter: UiParameter): boolean => {
  return isFieldFilterUiParameter(parameter) && parameter.fields.length > 0;
};
```

This way the type guard deals only with type-checking and `hasFields` deals with application logic.

Reply:
Ah, okay. Thanks for spelling it out, I see it is nicer to separate things and force callers of `hasFields` to also call `isFieldFilterUiParameter`. I've created a PR [here](https://github.com/metabase/metabase/pull/35207) with the change.